weighted MDS

doerlbh opened the pull request:
Enabled function for weighted MDS: pos2Dor3D = vis.weighted_mds(rdms, dim, weight)
Goal: Compute weighted multidimensional scaling (MDS) arrangement

Variables:
- rdms is the input RDM matrices to be approximated
- weight is the corresponding weights matrix (the same shape as rdms. if w is all ones, then the function performs unweighted MDS)
- dim defines the dimensionality of the target space

Notes:
- like rdms, the weighting matrices can be either 2d or 3d, to be processed in batches. Both are well supported.

doerlbh committed:
[FEAT] restructure util.vis modules

doerlbh committed:
[FEAT] add wMDS implementation

doerlbh committed:
[FEAT] include wmds function in the init of util

doerlbh committed:
[FEAT] update vis init

doerlbh committed:
[DBG] clean up variable naming to avoid confusion

doerlbh committed:
[FEAT] enable n-d MDS in vis

doerlbh committed:
[FEAT] add weighting factor into dim reduction function

doerlbh committed:
[FEAT] clean up notations, add 3d test case

doerlbh committed:
[FEAT] add batch weight matrix processing

doerlbh committed:
[FEAT] add batch weight matrix processing

doerlbh committed:
[FEAT] add test for wMDS functions with batched weights

doerlbh committed:
[FEAT] add test for wMDS functions with batched weights

doerlbh commented:
If possible, it would be great if this can be merged to the master before tomorrow, because there are Hackathon participants expressing interest in using it in their projects. Thank you!

JasperVanDenBosch commented on pyrsa/vis/mds_plot.py:
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
```
We don't have automatic api documentation anymore, so a new module (or renamed) requires adapting the narrative documentation (rst) as well

doerlbh replied:
Just included them. BTW, why can't I see the pylint etc. checks on my end any more? Is it just me, or everyone?

JasperVanDenBosch replied:
Not just you, it's been turned off (maybe something to discuss in the workshop?)

JasperVanDenBosch commented on pyrsa/vis/mds_plot.py:
```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+"""
+Definition of visualization methods
+
```
Since this doesn't contain a plot, and only transformation functions, maybe this should be clear in the name of the module. Perhaps there should even be a transformations sub package

doerlbh replied:
retitled inside to be more precise. BTW, I actually prefer mds_plot, since we have this question in an early PR (vis-dev) whether we should include exact MDS plotting etc., and the answer is not clear yet. Therefore, if additional plotting functions emerge, we can include them in mds_plot.py, similar to the conventions in rdm_plot.py and model_plot.py.

JasperVanDenBosch commented on pyrsa/vis/mds_plot.py:
```diff
@@ -0,0 +1,71 @@
+    return rdm_dimension_reduction(rdms, mds_emb, dim)
+
+
+def weighted_mds(rdms, dim=2, weight=None):
```
The two functions are very similar, have you considered to have just one? Then if weights are supplied you could create a `Weighted_MDS`.

doerlbh replied:
changed. thanks for the suggestion!

JasperVanDenBosch commented on tests/test_vis.py:
```diff
@@ -20,9 +20,40 @@ def test_vis_mds_output_shape_corresponds_to_inputs(self):
+                        dissimilarity_measure=mes,
+                        descriptors=des)
+        mds_emb = rsv.mds(rdms, dim=3)
+        self.assertEqual(mds_emb.shape, (8, 5, 3))
```
The tests only test the dimensions of the output. Ideally they would also test the difference in the weighted case

doerlbh replied:
done.

JasperVanDenBosch commented on pyrsa/util/__init__.py:
```diff
@@ -1 +1 @@
-from . import matrix
+from . import matrix
```
> missing newline at end of file

doerlbh replied:
fixed. thanks!

JasperVanDenBosch commented on pyrsa/vis/__init__.py:
```diff
@@ -1,3 +1,3 @@
 from .rdm_plot import show_rdm
-from .vis import mds,rdm_dimension_reduction
+from .mds_plot import mds,weighted_mds,rdm_dimension_reduction
```
> C0326: Exactly one space required after comma

doerlbh replied:
fixed. thanks!

JasperVanDenBosch reviewed (changes requested):
A few lint issues and I think it would be good to separate plotting code from transformations such as these. Also the tests currently only check for dimensions. Otherwise a good addition!

doerlbh committed:
[DBG] pylint fix

doerlbh committed:
[FEAT] combine two MDS into one method

doerlbh committed:
[FEAT] add rst docs for new modules

doerlbh committed:
[FEAT] add additional test in mds to test for differences

doerlbh commented:
Thanks @ilogue for the suggestion. All the requested changes are now included in the latest commits: additional docs, better styling, combine mds functions, module repurposing, and tests to compare differences between mds and wmds. 

Now the default function is: pos2Dor3D = vis.mds(rdms, dim=2, weight=None)

BTW, why can't I see the pylint etc. checks on my end any more? Is it just me, or everyone?

JasperVanDenBosch commented:
> BTW, why can't I see the pylint etc. checks on my end any more? Is it just me, or everyone?

No that's not just you, it's been turned off.

JasperVanDenBosch commented on docs/source/pyrsa.vis.mds_plot.rst:
```diff
@@ -0,0 +1,7 @@
+pyrsa.vis.colors module
+=======================
+
+.. automodule:: pyrsa.vis.colors
```
I think this should be `.. automodule:: pyrsa.vis.mds_plot`

doerlbh replied:
Oops, nice catch!. Just fixed it.

JasperVanDenBosch commented on pyrsa/util/vis_utils.py:
```diff
@@ -0,0 +1,450 @@
+from sklearn.metrics import euclidean_distances
+from sklearn.utils import check_random_state, check_array, check_symmetric
+from sklearn.externals.joblib import Parallel
+from sklearn.externals.joblib import delayed
```
We don't currently depend on Joblib yet, so we'd have to add this to the requirements file (currently this would fail with an import error I think)

doerlbh replied:
I thought if we already have sklearn as a requirement, wouldn't joblib be implicitly already also installed during pyrsa and sklearn installations?

JasperVanDenBosch replied:
You are right, I just tried and joblib installs along with sklearn. Hm I don't like joblib so much because it doesn't work well with out HPC cluster. It does some sneaky stuff that no-one fully understands. 
But I guess we have to accept it.

doerlbh replied:
I just updated the code to include joblib as a requirement. In addition, I removed two sklearn dependencies to lighten things up.

doerlbh replied:
Wait, so which one would you prefer? I just saw your comment popping up after my latest commit (there is some delay in Github). I guess we can also potentially go back to this issue later, to try to move away from joblib if people find it troublesome from time to time~

JasperVanDenBosch reviewed (changes requested):
Thanks for your edits. Sorry but I had not seen `vis_utils` yet.
It contains a lot of code based on `sklearn`. Does it make sense for us to maintain that? Is there a way around this? Like submit a PR with them?

Can someone else weigh in on this? @HeikoSchuett

doerlbh committed:
[DBG] fix typo in doc

doerlbh committed:
[DBG] fix typo in doc

doerlbh commented:
I just updated the code to include joblib as a requirement. In addition, I removed two sklearn function dependencies to lighten things up. Thanks for the careful read and considerations, @ilogue

doerlbh commented:
>  It contains a lot of code based on `sklearn`. Does it make sense for us to maintain that? Is there a way around this? Like submit a PR with them?

My personal take is that, for the current stage (and semi-public Hackathon tomorrow), we can first maintain it lightly on our end. In the long term, I can submit a PR at sklearn but it might take longer for them to review, approve and release etc.. What do you think?

doerlbh committed:
[FEAT] add requirements and simplify sklearn dependencies

JasperVanDenBosch commented:
> My personal take is that, for the current stage (and semi-public Hackathon tomorrow), we can first maintain it lightly on our end. In the long term, I can submit a PR at sklearn but it might take longer for them to review, approve and release etc.. What do you think?

That sounds reasonable. Could you open an issue here on the pyrsa repo, to remind us to deal with the `sklearn` code later?

JasperVanDenBosch commented:
I'm getting a few errors when running the tests, is this just me:

```
======================================================================
ERROR: test_vis_3d_mds_output_shape_corresponds_to_inputs (tests.test_vis.TestVIS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/my_code_dir/pyrsa/tests/test_vis.py", line 33, in test_vis_3d_mds_output_shape_corresponds_to_inputs
    mds_emb = rsv.mds(rdms, dim=3)
  File "/my_code_dir/pyrsa/pyrsa/vis/mds_plot.py", line 61, in mds
    return rdm_dimension_reduction(rdms, mds_emb, dim, weight)
  File "/my_code_dir/pyrsa/pyrsa/vis/mds_plot.py", line 33, in rdm_dimension_reduction
    ws = weight_to_matrices(weight)
  File "/my_code_dir/pyrsa/pyrsa/util/vis_utils.py", line 39, in weight_to_matrices
    if x.ndim == 2:
AttributeError: 'NoneType' object has no attribute 'ndim'

======================================================================
ERROR: test_vis_3d_weighted_mds_output_behaves_like_mds (tests.test_vis.TestVIS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/my_code_dir/pyrsa/tests/test_vis.py", line 72, in test_vis_3d_weighted_mds_output_behaves_like_mds
    wes = np.random.ones((8,10))
AttributeError: module 'numpy.random' has no attribute 'ones'

======================================================================
ERROR: test_vis_mds_output_shape_corresponds_to_inputs (tests.test_vis.TestVIS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/my_code_dir/pyrsa/tests/test_vis.py", line 23, in test_vis_mds_output_shape_corresponds_to_inputs
    mds_emb = rsv.mds(rdms)
  File "/my_code_dir/pyrsa/pyrsa/vis/mds_plot.py", line 61, in mds
    return rdm_dimension_reduction(rdms, mds_emb, dim, weight)
  File "/my_code_dir/pyrsa/pyrsa/vis/mds_plot.py", line 33, in rdm_dimension_reduction
    ws = weight_to_matrices(weight)
  File "/my_code_dir/pyrsa/pyrsa/util/vis_utils.py", line 39, in weight_to_matrices
    if x.ndim == 2:
AttributeError: 'NoneType' object has no attribute 'ndim'

======================================================================
ERROR: test_vis_weighted_mds_output_behaves_like_mds (tests.test_vis.TestVIS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/my_code_dir/pyrsa/tests/test_vis.py", line 60, in test_vis_weighted_mds_output_behaves_like_mds
    wes = np.random.ones((8,10))
AttributeError: module 'numpy.random' has no attribute 'ones'

----------------------------------------------------------------------
```

doerlbh committed:
[DBG] fix bug in vis test

doerlbh committed:
[DBG] accommodate Nonetype weight

doerlbh commented:
> I'm getting a few errors when running the tests, is this just me:

Just fixed. 

> That sounds reasonable. Could you open an issue here on the pyrsa repo, to remind us to deal with the `sklearn` code later?

Done.

doerlbh commented:
> That sounds reasonable. Could you open an issue here on the pyrsa repo, to remind us to deal with the `sklearn` code later?

I also already submitted a PR at sklearn as well, which it should take quite a long time (as there are >700 PRs before mine). BTW, their (sklearn's) pylint and pytest modules are working properly -- it's weird that ours (pyrsa's) isn't.

JasperVanDenBosch commented:
> I also already submitted a PR at sklearn as well, which it should take quite a long time (as there are >700 PRs before mine). 

Excellent work!

> BTW, their (sklearn's) pylint and pytest modules are working properly -- it's weird that ours (pyrsa's) isn't.

Yeah if I remember correctly some people found the rules too strict so we turned it off. Perhaps we should discuss this at the workshop

doerlbh commented:
> Yeah if I remember correctly some people found the rules too strict so we turned it off. Perhaps we should discuss this at the workshop

Good idea!

doerlbh committed:
[DBG] improve lint based on sklearn tests

doerlbh committed:
[DBG] update comment for docs API

doerlbh committed:
[DBG] update comment for docs API

doerlbh committed:
[DBG] update description of wMDS
